packages-exp/auth-exp/src/core/credentials/oauth.ts

```diff
@@ -0,0 +1,161 @@
+  accessToken?: string | null;
+  oauthToken?: string;
+  secret?: string;
+  oauthTokenSecret?: string;
+  nonce?: string;
+  pendingToken?: string;
+  providerId: externs.ProviderId;
+  signInMethod: externs.SignInMethod;
+}
+
+export class OAuthCredential
+  implements externs.OAuthCredential, AuthCredential {
+  idToken?: string;
+  accessToken?: string;
+  secret?: string;
+  nonce?: string;
+  private pendingToken: string | null = null;
+
+  private constructor(
+    readonly providerId: externs.ProviderId,
+    readonly signInMethod: externs.SignInMethod
+  ) {}
+
+  static _fromParams(params: OAuthCredentialParams): OAuthCredential {
+    const cred = new OAuthCredential(params.providerId, params.signInMethod);
+
+    if (params.idToken || params.accessToken) {
+      // OAuth 2 and either ID token or access token.
+      if (params.idToken) {
+        cred.idToken = params.idToken;
+      }
+
+      if (params.accessToken) {
+        cred.accessToken = params.accessToken;
+      }
+
+      // Add nonce if available and no pendingToken is present.
+      if (params.nonce && !params.pendingToken) {
+        cred.nonce = params.nonce;
+      }
+
+      if (params.pendingToken) {
+        cred.pendingToken = params.pendingToken;
+      }
+    } else if (params.oauthToken && params.oauthTokenSecret) {
+      // OAuth 1 and OAuth token with token secret
+      cred.accessToken = params.oauthToken;
+      cred.secret = params.oauthTokenSecret;
+    } else {
+      fail('todo', AuthErrorCode.ARGUMENT_ERROR);
+    }
+
+    return cred;
+  }
+
+  toJSON(): object {
+    return {
+      idToken: this.idToken,
+      accessToken: this.accessToken,
+      secret: this.secret,
+      nonce: this.nonce,
+      pendingToken: this.pendingToken,
+      providerId: this.providerId,
+      signInMethod: this.signInMethod
+    };
+  }
+
+  static fromJSON(obj: object): externs.OAuthCredential | null {
+    const { providerId, signInMethod, ...rest }: Partial<OAuthCredential> = obj;
+    if (!providerId || !signInMethod) {
+      return null;
+    }
+
+    const cred = new OAuthCredential(providerId, signInMethod);
+    Object.assign(cred, rest);
+    return cred;
+  }
+
+  _getIdTokenResponse(auth: Auth): Promise<IdTokenResponse> {
+    const request = this.makeRequest();
+    return signInWithIdp(auth, request);
+  }
+
+  _linkToIdToken(auth: Auth, idToken: string): Promise<IdTokenResponse> {
+    const request = this.makeRequest();
+    request.idToken = idToken;
+    return signInWithIdp(auth, request);
+  }
+
+  _getReauthenticationResolver(auth: Auth): Promise<IdTokenResponse> {
+    const request = this.makeRequest();
+    request.autoCreate = false;
+    return signInWithIdp(auth, request);
+  }
+
+  private makeRequest(): SignInWithIdpRequest {
+    const request: SignInWithIdpRequest = {
+      requestUri: IDP_REQUEST_URI,
+      returnSecureToken: true,
+      postBody: null
+    };
+
+    const postBody: { [key: string]: string } = {};
```

Comment:
nit: consider creating a type here so you don't have to ['field'] all over the place, alternatively this can be a Record